### PR TITLE
feat: implement strategic merge patching for API server admission config

### LINKED ISSUE #5895: Strategic merge for admission control config
Admission control API server config should support merging like `network: interfaces`. 

E.g. updating the Pod Security configuraiton.

### PR DESCRIPTION
**Pull request by smira**
The testcase explains it better, but tl;dr is that this allows to do
strategic merge patching e.g. for the Pod Security configuration.

Fixes #5895

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

### CODE REVIEW

File: pkg/machinery/config/merge/merge.go [hunk 1]
```diff
@@ -64,6 +69,8 @@ func merge(vl, vr reflect.Value, replace bool) error {
 			return nil
 		}
 
+		log.Printf("dereference %v -> %v", vl, vl.Elem())
```
[the remark this answers is not in the document]
Comment (smira) [reply pkg/machinery/config/merge/merge.go:72]: oops... need to clean up this :)

**Review by frezbo — APPROVED** [review]

File: pkg/machinery/config/merge/merge.go [hunk 2]
```diff
@@ -102,11 +106,32 @@ func merge(vl, vr reflect.Value, replace bool) error {
...
-				v.Set(vl.MapIndex(k))
+				valueType := tl.Elem()
+
+				var v, rightV reflect.Value
+
+				if valueType.Kind() == reflect.Interface { // special case for map[string]interface{}
+					// here we have to instantiate the actual type behind in the `interface{}`, otherwise it's not settable
+					valueType = vl.MapIndex(k).Elem().Type()
+
+					if valueType != vl.MapIndex(k).Elem().Type() {
```
Comment (DmitriyMV) [inline pkg/machinery/config/merge/merge.go:117]: Isn't valueType set from vl.MapIndex(k).Elem().Type()?

File: pkg/machinery/config/merge/merge.go [hunk 3]
```diff
@@ -102,11 +106,32 @@ func merge(vl, vr reflect.Value, replace bool) error {
...
 			if vl.MapIndex(k) != zeroValue {
-				v := reflect.New(tl.Elem()).Elem()
-				v.Set(vl.MapIndex(k))
+				valueType := tl.Elem()
+
+				var v, rightV reflect.Value
+
+				if valueType.Kind() == reflect.Interface { // special case for map[string]interface{}
+					// here we have to instantiate the actual type behind in the `interface{}`, otherwise it's not settable
+					valueType = vl.MapIndex(k).Elem().Type()
```
Comment (DmitriyMV) [inline pkg/machinery/config/merge/merge.go:115]: Can we extract `vl.MapIndex(k).Elem()` into the variable?

File: pkg/machinery/config/merge/merge.go (hunk 2 above)
Comment (smira) [reply pkg/machinery/config/merge/merge.go:117]: oops, it should have been `vr` there, nice catch

File: pkg/machinery/config/merge/merge.go (hunk 3 above)
Comment (smira) [reply pkg/machinery/config/merge/merge.go:115]: I don't think we need it, as it should have been `vr` above, just checking that left and right are of the same type

File: pkg/machinery/config/merge/merge.go [hunk 4]
```diff
@@ -102,11 +106,32 @@ func merge(vl, vr reflect.Value, replace bool) error {
...
+
+					if valueType != vr.MapIndex(k).Elem().Type() {
+						return fmt.Errorf("merge type mismatch left %v right %v", valueType, vl.MapIndex(k).Elem().Type())
```
Comment (DmitriyMV) [inline pkg/machinery/config/merge/merge.go:118]:
```suggestion
						return fmt.Errorf("merge type mismatch left %v right %v", valueType, vr.MapIndex(k).Elem().Type())
```

Probably?

Comment (smira) [reply pkg/machinery/config/merge/merge.go:118]: ok, I'm blind :eyes: thanks

**Review by sauterp — APPROVED** [review]

**Pull request merged**